### Software tools developed specifically for MinION sequence data <sup>[1](https://doi.org/10.1186/s13059-016-1103-0)</sup>

| Name | Applications | Link |
| ------ | ------ | ------ |
| Poretools | Sequence data extraction and statistics | https://github.com/arq5x/poretools |
| poRe | Sequence extraction and basic statistics | https://sourceforge.net/projects/rpore/ |
| BWA MEM | Sequence alignment | https://github.com/lh3/bwa |
| LAST | Sequence alignment | http://last.cbrc.jp/ |
| NanoOK | Sequence alignment, statistics, and visualization | https://documentation.tgac.ac.uk/display/NANOOK/ |
| marginAlign | Sequence alignment, SNV calling, and statistics | https://github.com/benedictpaten/marginAlign |
| Nanopolish | Signal alignment and SNV calling | https://github.com/jts/nanopolish |
| GraphMap | Sequence alignment and SNV calling | https://github.com/isovic/graphmap |
| minimap | Fast approximate mapping | https://github.com/lh3/minimap |
| miniasm | De novo assembly | https://github.com/lh3/miniasm |
| CANU | De novo assembly | https://github.com/marbl/canu |
| Nanocorrect | De novo assembly | https://github.com/jts/nanocorrect |
| PoreSeq | De novo assembly and SNV calling | https://github.com/tszalay/poreseq |
| NaS | De novo assembly | https://github.com/institut-de-genomique/NaS |
| Nanocorr | De novo assembly | https://github.com/jgurtowski/nanocorr |
| Mash | Species identification and fast approximate alignments | https://github.com/marbl/mash |
| minoTour | Real-time data analysis | https://github.com/minoTour/minoTour |
| Read Until | Selective sequencing | https://github.com/mattloose/RUscripts |
| Nanocall | Local base-calling | https://github.com/mateidavid/nanocall |
| DeepNano | Recurrent neural network (RNN)-based base-calling | https://bitbucket.org/vboza/deepnano |
